knimin/lib/data_access.py

```diff
@@ -3532,6 +3533,204 @@ def read_dna_plate(self, dna_plate_id):
+            corresponding to the normalized water and sample volumes
+        """
+        with TRN:
+            # get plate details
+            sql = """SELECT *
+                     FROM pm.shotgun_normalized_plate
+                     WHERE shotgun_normalized_plate_id = %s"""
+            TRN.add(sql, [shotgun_normalized_plate_id])
+            res = TRN.execute_fetchindex()
+            if not res:
+                raise ValueError("normalized shotgun plate %s does not exist" %
+                                 shotgun_normalized_plate_id)
+
+            res = dict(res[0])
+            res.pop('shotgun_normalized_plate_id')  # it's redundant
+
+            # get plate shape
+            sql = """SELECT rows, cols
+                     FROM pm.shotgun_plate JOIN
+                          pm.plate_type USING(plate_type_id)
+                     WHERE shotgun_plate_id = %s"""
+            TRN.add(sql, [res['shotgun_plate_id']])
+            plate_shape = dict(TRN.execute_fetchindex()[0])
+            n_rows = plate_shape['rows']
+            n_cols = plate_shape['cols']
+            sample_volumes = np.zeros((n_rows, n_cols))
+            water_volumes = np.zeros((n_rows, n_cols))
+
+            # get well values
+            sql = """SELECT row, col, sample_volume_nl, water_volume_nl
+                     FROM pm.shotgun_normalized_plate_well_values
+                     WHERE shotgun_normalized_plate_id = %s"""
+            TRN.add(sql, [shotgun_normalized_plate_id])
+            well_values = TRN.execute_fetchindex()
+            for row, col, sample_volume_nl, water_volume_nl in well_values:
+                sample_volumes[row, col] = sample_volume_nl
+                water_volumes[row, col] = water_volume_nl
+            res['plate_normalization_water'] = water_volumes
+            res['plate_normalization_sample'] = sample_volumes
+
+        return res
+
+    def delete_normalized_shotgun_plate(self, shotgun_normalized_plate_id):
+        """Delete a normalized shotgun plate
+
+        Parameters
+        ----------
+        shotgun_normalized_plate_id : int
+            The ID of the normalized shotgun plate to delete.
+
+        Raises
+        ------
+        ValueError
+            If the normalized_shotgun_plate_id does not exist
+        """
+        ### what mechanisms exist to prohibit the execution of this method for
+        ### an already sequenced set of plates?
+        with TRN:
+            # get plate details
+            sql = """SELECT *
+                     FROM pm.shotgun_normalized_plate
+                     WHERE shotgun_normalized_plate_id = %s"""
+            TRN.add(sql, [shotgun_normalized_plate_id])
+            res = TRN.execute_fetchindex()
+            if not res:
+                raise ValueError("normalized shotgun plate %s does not exist" %
+                                 shotgun_normalized_plate_id)
+
+            sql = """DELETE FROM pm.shotgun_normalized_plate_well_values
```

Comment:
Minor, but you can avoid making the SQL query in line 3716 by modifying this SQL query to return the shotgun_normalized_plate_id. No need to change if you don't feel like but I wanted to mention it.